File: pipeline-work/scripts/validate_against_vocab.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# Validate every cert YAML against its per-R schema.
#
# For each yaml/R<NN>/<edition>/<cert>.yaml:
#   1. Load schema/R<NN>.yaml
#   2. For each aspect in scope.aspects with an enum constraint, check the
#      cert's value(s) against the allowed set.
#   3. Collect violations and print a per-R summary + sample violations.

require "yaml"
require "pathname"

ROOT = Pathname.new(File.expand_path("..", __dir__))
YAML_ROOT = ROOT + "yaml"
SCHEMA_DIR = ROOT + "schema"

def load_r_schema(r_str)
  path = SCHEMA_DIR + "#{r_str}.yaml"
  return nil unless path.exist?
  YAML.load_file(path)
end

def values_to_array(v)
  return [] if v.nil?
  return [v] unless v.is_a?(Array)
  v
end

def validate_cert(yaml_path, schema)
  data = YAML.load_file(yaml_path)
  return [{ issue: "no_yaml_data" }] unless data.is_a?(Hash)

  r_str = yaml_path.parent.parent.basename.to_s
  chars = data.dig("characteristics", "type_level") || {}
  aspects = schema.dig("scope", "aspects") || {}

  violations = []
  aspects.each do |label, spec|
    next unless spec.is_a?(Hash)
    type = spec["type"]
    next unless %w[enum enum_multi].include?(type)
    allowed = spec["values"] || []
    next if allowed.empty?
    next unless chars.key?(label)

    value_obj = chars[label]
    next unless value_obj.is_a?(Hash)
    raw = value_obj["value"]
    vals = values_to_array(raw).map { |v| v.is_a?(Numeric) ? v.to_s : v.to_s.strip }
    bad = vals.reject { |v| allowed.include?(v) }
    next if bad.empty?
    violations << { label: label, bad: bad, allowed: allowed }
  end
  violations
end

# Walk every cert
puts "=== VALIDATION ==="
per_r = Hash.new { |h, k| h[k] = { total: 0, violations: 0, samples: [] } }
all_violations = []

Pathname.glob(YAML_ROOT + "R*" + "*" + "*.yaml").sort.each do |yp|
  r_str = yp.parent.parent.basename.to_s
  per_r[r_str][:total] += 1
  schema = load_r_schema(r_str)
  unless schema
    per_r[r_str][:violations] += 1
    per_r[r_str][:samples] << { cert: yp.basename.to_s, issue: "no_schema" }
    next
  end
  vs = validate_cert(yp, schema)
  next if vs.empty?
  per_r[r_str][:violations] += vs.size
  vs.first(2).each { |v| per_r[r_str][:samples] << { cert: yp.basename.to_s, **v } }
  all_violations.concat(vs.map { |v| { R: r_str, cert: yp.basename.to_s, **v } })
end

puts
puts "=== PER-R SUMMARY ==="
puts format("%-6s %8s %12s", "R", "certs", "violations")
per_r.keys.sort_by { |r| r[1..].to_i }.each do |r|
  v = per_r[r]
  flag = v[:violations].zero? ? "✓" : "✗"
  puts format("  %s %-4s %8d %12d", flag, r, v[:total], v[:violations])
end

puts
puts "=== SAMPLE VIOLATIONS (first 20) ==="
all_violations.first(20).each do |v|
  puts "  [#{v[:R]}] #{v[:cert]}"
  if v[:issue]
    puts "      issue: #{v[:issue]}"
  else
    puts "      #{v[:label]}: #{v[:bad].inspect} not in #{v[:allowed].inspect}"
  end
end

puts
puts "Total violations: #{all_violations.size}"

File: pipeline-work/scripts/validate_against_vocab.py
#!/usr/bin/env python3
"""Validate all cert YAMLs against their per-R vocabulary.

For each cert:
  1. Look up its R-number's vocabulary (schema/vocabularies/R<NN>.yaml).
  2. For each declared characteristic with an enum, check the cert's value(s)
     against the allowed set.
  3. Report violations: out-of-enum values, missing R-vocab, etc.

Outputs a per-R summary + a list of every violation for review.
"""
from __future__ import annotations

import sys
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

import yaml  # noqa: E402


def load_vocab(r_str: str) -> dict | None:
    path = ROOT / "schema" / "vocabularies" / f"{r_str}.yaml"
    if not path.exists():
        return None
    return yaml.safe_load(path.read_text())


def check_enum(values, allowed: list[str]) -> list[str]:
    """Return list of values not in allowed set."""
    if isinstance(values, str):
        values = [values]
    elif not isinstance(values, list):
        values = [str(values)]
    return [v for v in values if v not in allowed and v is not None]


def validate_cert(yaml_path: Path) -> list[dict]:
    """Return list of violations for one cert."""
    data = yaml.safe_load(yaml_path.read_text())
    r_str = yaml_path.parts[-3]
    vocab = load_vocab(r_str)
    if vocab is None:
        return [{"cert": yaml_path.stem, "R": r_str, "issue": "no_vocab_file"}]

    violations: list[dict] = []
    char_specs = vocab.get("characteristics") or {}
    chars = (data.get("characteristics") or {}).get("type_level") or {}

    for label, spec in char_specs.items():
        if not spec or spec.get("type") not in ("enum", "enum_multi"):
            continue
        allowed = spec.get("values") or []
        if not allowed:
            continue
        if label not in chars:
            continue
        value_obj = chars[label]
        if not isinstance(value_obj, dict):
            continue
        value = value_obj.get("value")
        if value is None:
            continue
        bad = check_enum(value, allowed)
        if bad:
            violations.append({
                "cert": yaml_path.stem,
                "R": r_str,
                "label": label,
                "bad_values": bad,
                "allowed": allowed,
            })
    return violations


def main() -> int:
    yaml_root = ROOT / "yaml"
    all_violations: list[dict] = []
    per_r = defaultdict(lambda: {"total": 0, "violations": 0, "samples": []})

    for yp in sorted(yaml_root.glob("R*/*/*.yaml")):
        r_str = yp.parts[-3]
        per_r[r_str]["total"] += 1
        vs = validate_cert(yp)
        if vs:
            per_r[r_str]["violations"] += len(vs)
            all_violations.extend(vs)
            per_r[r_str]["samples"].extend(vs[:2])

    print(f"\n=== PER-R SUMMARY ===")
    print(f"{'R':6s} {'certs':>6s} {'violations':>11s}")
    for r in sorted(per_r.keys(), key=lambda k: int(k[1:])):
        v = per_r[r]
        flag = "✓" if v["violations"] == 0 else "✗"
        print(f"  {flag} {r:4s} {v['total']:>6d} {v['violations']:>11d}")

    print(f"\n=== SAMPLE VIOLATIONS (first 30) ===")
    for v in all_violations[:30]:
        print(f"  [{v['R']}] {v['cert']}")
        if "issue" in v:
            print(f"      issue: {v['issue']}")
        else:
            print(f"      {v['label']}: {v['bad_values']} not in {v['allowed']}")

    print(f"\nTotal violations: {len(all_violations)}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
